pkg/auth/authenticator/password/ldappassword/util/identityfactory.go

```diff
@@ -0,0 +1,60 @@
+package util
+
+import (
+	"fmt"
+
+	"github.com/go-ldap/ldap"
+
+	authapi "github.com/openshift/origin/pkg/auth/api"
+)
+
+// LDAPUserIdentityFactory creates Identites for LDAP user entries.
+type LDAPUserIdentityFactory interface {
+	IdentityFor(user *ldap.Entry) (identity authapi.UserIdentityInfo, err error)
+}
+
+// LDAPUserToIdentityMapping holds LDAP attributes that map to OpenShift Identity fields in oder to
+// create a deterministic mapping from LDAP user entries to OpenShift Identities
+type LDAPUserToIdentityMapping struct {
+	// EmailAttribute is the optional list of LDAP attributes to use for the email address of the user
+	// identity. The first attribute with a non-empty value is used.
+	EmailAttribute []string
+	// NameAttribute is the optional list of LDAP attributes to use for the display name of the user
+	// identity. The first attribute with a non-empty value is used.
+	NameAttribute []string
+	// PreferredUsernameAttribute is the optional list of LDAP attributes to use for the preferred
+	// username of the user identity. The first attribute with a non-empty value is used. If not
+	// specified, the id determined by AttributeID is used as the preferred login.
+	PreferredUsernameAttribute []string
+	// IDAttribute is the required list of LDAP attributes to use for the id address of the user identity.
+	// The first attribute with a non-empty value is used. If no attributes have values, login fails.
+	IDAttribute []string
+}
+
+// DefaultLDAPUserIdentityFactory creates Identities for LDAP user entries using an LDAPUserToIdentityMapping
+type DefaultLDAPUserIdentityFactory struct {
+	ProviderName string
+	Mapping      *LDAPUserToIdentityMapping
```

Comment:
Why a pointer?  It's a small object that we probably don't want to mutate in a non-obvious part of the code.